### ISSUE TITLE
Overload decorator lines classified as Unanalyzed

### ISSUE THREAD

**Issue by nsoulat** [python/mypy#21638]
**Bug Report**

`@overload` decorator lines are classified as `Unanalyzed` in `--lineprecision-report`. The lineprecision ratio `(Precise + Imprecise) / (Lines - Empty)` therefore drops with every overload added, even when all signatures are fully annotated.

**To Reproduce**

```python
# mre.py
from typing import Literal, overload


class C:
    @overload
    def f(self, x: Literal["a"]) -> str: ...
    @overload
    def f(self, x: Literal["b"]) -> int: ...
    def f(self, x: Literal["a", "b"]) -> str | int:
        return "hello" if x == "a" else 42
```

```
$ mypy mre.py --lineprecision-report report/
Success: no issues found in 1 source file
$ cat report/lineprecision.txt
Name  Lines  Precise  Imprecise  Any  Empty  Unanalyzed
-------------------------------------------------------
mre      10        6          0    0      2           2
```

**Expected Behavior**

Every line is fully annotated and mypy reports no errors, so the ratio should be `8/8 = 100%`. The 2 `Unanalyzed` lines are the two `@overload` decorator lines. `overload` is a well-typed symbol from `typing`; those lines should be `Precise` (or at minimum `Empty`), treating `@overload` as a pure annotation marker with no runtime value.

**Actual Behavior**

`StatisticsVisitor.process_node` looks up each expression in `self.typemap`. The `NameExpr` for `@overload` is not populated in the typemap (mypy treats it as a special form during semantic analysis), so `typemap.get(node)` returns `None` → `type(None)` → `record_line(TYPE_UNANALYZED)`.

Each overload variant therefore contributes 1 `Precise` line (the `def`) and 1 `Unanalyzed` line (`@overload`). Adding more fully-annotated overloads strictly lowers the ratio, making it impossible to hold a stable lineprecision gate on any file with a growing overload set.

There is a related `# TODO` in the same file: `pass  # TODO: Handle overloaded functions, etc.` (in `record_call_target_precision`).

**Your Environment**

- Mypy version used: 2.1.0
- Mypy command-line flags: `--lineprecision-report`
- Python version used: 3.13

**Comment by Lohitha0-0** [comment]
Hi, I'd like to work on this!

I've traced the issue to `StatisticsVisitor.process_node` in `mypy/stats.py` — the `NameExpr` for `@overload` isn't present in the typemap since mypy treats it as a special form during semantic analysis, causing it to fall through to `TYPE_UNANALYZED`. The fix would be to special-case `@overload` decorators and classify them as `Precise` instead.

Could someone assign this to me? Happy to submit a PR.

**Comment by gregoiredx** [comment]
@Lohitha0-0 have you looked at @nsoulat PR https://github.com/python/mypy/pull/21642 ?

Does it look like the good approach to you as well?

**Comment by Lohitha0-0** [comment]
@gregoiredx Yeah, I think this is the right approach. It reuses record_precise_if_checked_scope — the same function already handling None/True/False/Ellipsis — rather than adding a bespoke branch, so it fits existing precedent in the file.
It's also scope-aware, which the new tests prove: @overload in a checked scope counts toward Precise, while in an unchecked scope it falls through to Any instead, matching how every other line in an unchecked scope is already reported. That's more correct than just hardcoding it to always be Precise. It also uses OVERLOAD_NAMES rather than a string literal, so it covers typing_extensions.overload too.
On the "spirit of the report" question — the issue asked for Precise "or at minimum Empty." This goes with Precise, but that mirrors how the file already treats other zero-runtime-value markers, so it's extending an existing pattern rather than introducing a new inconsistency.
+1 from me, looks good to merge.

**Comment by gregoiredx** [comment]
Ok cool, let's go! Can you merge it?

**Comment by Lohitha0-0** [comment]
I don't have merge rights on this repo, so it'd need one of the mypy maintainers to approve and merge. Might be worth tagging one directly, or @nsoulat could ping in the PR thread.
